migration/idempotent/001-base.sql

```diff
@@ -1106,36 +1106,40 @@ CREATE OR REPLACE FUNCTION _prom_catalog.get_or_create_label_ids(metric_name TEX
     VOLATILE
     SET search_path = pg_catalog, pg_temp
 AS $$
-        WITH cte as (
-        SELECT
-            -- only call the functions to create new key positions
-            -- and label ids if they don't exist (for performance reasons)
-            lkp.pos as known_pos,
-            coalesce(l.id, _prom_catalog.get_or_create_label_id(kv.key, kv.value)) label_id,
-            kv.key key_str,
-            kv.value val_str
-        FROM ROWS FROM(unnest(label_keys), UNNEST(label_values)) AS kv(key, value)
-            LEFT JOIN _prom_catalog.label l
-               ON (l.key = kv.key AND l.value = kv.value)
-            LEFT JOIN _prom_catalog.label_key_position lkp ON
-            (
-                    lkp.metric_name = get_or_create_label_ids.metric_name AND
-                    lkp.key = kv.key
+    BEGIN
+        RETURN QUERY EXECUTE format($query$
```

Comment:
I know that there is a lot of "black magic" around the query planner and optimising queries, but I have a few issues:

1. The fact that we've restructured this code for a specific outcome is not evident either in the code (e.g. as a comment), or in the tests (to assert that we're actually getting the desired query plan).
2. Based on the description of the commit, I don't understand why this works, or why this is the best way to achieve the desired outcome.

Reply:
1.  While testing I've tried turning off generic plan `set plan_cache_mode=force_custom_plan;` and also forcing planner to not use sequential scan `set enable_seqscan=off`, however non of that helped 🤷 . It seems that somehow planner gets confused when planning a query that uses arguments. The PR description explains the difference about query plans. If you feel strongly about it I could add code comment about it as well?
Making tests for function plan is a bit tricky. The `EXPLAIN` of a function call does not show inner query plan (like it does for a normal SQL query). The only way (that I am aware of) to get it is to use `auto_explain` library that will print execution plan to PG log...so it requires quite a bit of work...
2. Not really sure what part is confusing. I shared two plans - one that does sequential scan and another that does the index scan. Index scan is almost always preferred (unless there is a very little data to scan). Our benchmarks are showing from 10X-1000X improvement (depending on number of labels) for this fix.
What else do you feel we should add to make it more clear?